### PR TITLE
85 compile model

### PR DESCRIPTION
**Pull request by kirilklein**

### CODE REVIEW

**Comment by kirilklein** [comment]
Pretrain compilation works.
Finetuning compilation fails likely due to packing/unpacking sequences in the bigru head.
Trying not to compile the head for now. Alternatively, need to adjust head, e.g. removal of packing/unpacking and compare speed.

**Comment by kirilklein** [comment]
Compilation works for PT and FT.
However no speedup is observed for FT
Tried the following:

- increase num workers in dataloader + pin_memory
- introduce bucketed padding
- Setting explicitly to assume static by default

**Comment by kirilklein** [comment]
Profiling output:
| **Name**                                                    | **Self CPU %** | **Self CPU** | **CPU total %** | **CPU total** | **CPU time avg** | **Self CUDA** | **Self CUDA %** | **CUDA total** | **CUDA time avg** | **CPU Mem** | **Self CPU Mem** | **CUDA Mem** | **Self CUDA Mem** | **# of Calls** |
|:------------------------------------------------------------|:--------------:|:------------:|:---------------:|:-------------:|:----------------:|:-------------:|:---------------:|:--------------:|:-----------------:|:----------:|:---------------:|:-----------:|:-----------------:|--------------:|
| **aten::bmm**                                              | 0.12%          | 21.182ms     | 0.16%           | 27.581ms      | 30.645us         | 5.990s        | 37.89%          | 5.990s        | 6.655ms          | 0 b        | 0 b             | 1373.52 Gb  | 1373.52 Gb       | 900          |
| **ampere_sgemm_128x128_nn**                                | 0.00%          | 0.000us      | 0.00%           | 0.000us       | 0.000us          | 5.989s        | 37.88%          | 5.989s        | 9.982ms          | 0 b        | 0 b             | 0 b         | 0 b              | 600          |
| **aten::add**                                              | 0.12%          | 20.697ms     | 0.20%           | 34.029ms      | 20.017us         | 2.461s        | 15.57%          | 2.461s        | 1.448ms          | 0 b        | 0 b             | 1512.26 Gb  | 1512.26 Gb       | 1700         |
| **void at::native::elementwise_kernel<128, 2, at::nati...** | 0.00%          | 0.000us      | 0.00%           | 0.000us       | 0.000us          | 2.304s        | 14.58%          | 2.304s        | 1.163ms          | 0 b        | 0 b             | 0 b         | 0 b              | 1982         |
| **aten::mul**                                              | 0.15%          | 25.700ms     | 0.24%           | 42.158ms      | 19.163us         | 1.956s        | 12.37%          | 1.956s        | 889.067us        | 0 b        | 0 b             | 1498.54 Gb  | 1498.54 Gb       | 2200         |
| **aten::_softmax**                                         | 0.02%          | 3.121ms      | 0.03%           | 5.856ms       | 19.520us         | 1.720s        | 10.88%          | 1.720s        | 5.733ms          | 0 b        | 0 b             | 1331.89 Gb  | 1331.89 Gb       | 300          |
| **void (anonymous namespace)::softmax_warp_forward<flo...** | 0.00%          | 0.000us      | 0.00%           | 0.000us       | 0.000us          | 1.720s        | 10.88%          | 1.720s        | 2.896ms          | 0 b        | 0 b             | 0 b         | 0 b              | 594          |
| **void at::native::vectorized_elementwise_kernel<4, at...** | 0.00%          | 0.000us      | 0.00%           | 0.000us       | 0.000us          | 1.704s        | 10.78%          | 1.704s        | 905.193us        | 0 b        | 0 b             | 0 b         | 0 b              | 1882         |
| **aten::mm**                                               | 0.19%          | 33.227ms     | 0.26%           | 45.775ms      | 32.697us         | 1.319s        | 8.34%           | 1.319s        | 942.106us        | 0 b        | 0 b             | 180.36 Gb   | 180.36 Gb        | 1400         |
| **aten::copy_**                                            | 0.53%          | 92.777ms     | 83.24%          | 14.475s       | 708.181us        | 629.260ms     | 3.98%           | 629.260ms     | 30.787us         | 0 b        | 0 b             | 0 b         | 0 b              | 20439        |

**Self CPU time total:** 17.388s  
**Self CUDA time total:** 15.809s

**Comment by kirilklein** [comment]
A lot of time is spend on copy

**Comment by kirilklein** [comment]
Try the following:

- set pin_memory_device to GPU
- increase prefetch factor
- set non_blocking to true
- reduce GPU logging frequency.

Further things to try:

- compute metrics on GPU (especially for fine-tuning where logits are moved)

**Comment by kirilklein** [comment]
Split PR into:

- select_cohort:
  - setup for azure
  - fix argname (age)  
- new black + improve readme
- pass more params to dataloader with improved defaults (e.g. higher number of num_workers + persist_memory)
- improved metric computation (move only metric result not logits between devices)
- compilation

**Comment by kirilklein** [comment]
> Try the following:
> 
> * set pin_memory_device to GPU
> * increase prefetch factor
> * set non_blocking to true
> * reduce GPU logging frequency.
> 
> Further things to try:
> 
> * compute metrics on GPU (especially for fine-tuning where logits are moved)

- No speedup from increasing num_worker or setting pin_memory to true
- no speed up for compiled model for pt or ft
